### PR TITLE
Pluto - Press and Hold Button

### PR DESCRIPTION
**Pull request by emilbon99**
# Feature Pull Request Template

## Key Information

- **Linear Issue**: [SY-1394](https://linear.app/synnax/issue/SY-1394/button-press-and-hold)

## Description

Adds the ability to set the `delay` prop on a button, which makes sure the user holds the button for a certain duration before it activates. Also added the ability to set a button delay in the console.

## Basic Readiness

- [x] I have performed a self-review of my code.
- [x] I have added relevant tests to cover the changes to CI.
- [x] I have needed QA steps to the [release
      candidate](/.github/PULL_REQUEST_TEMPLATE/rc.md) template that cover these changes.
- [x] I have updated in-code documentation to reflect the changes.
- [x] I have updated user-facing documentation to reflect the changes.

## Backwards Compatibility

The following makes sure that this feature does not break backwards compatability.

### Data Structures

- [x] Server - I have ensured that previous versions of stored data structures are
      properly migrated to new formats.
- [x] Console - I have ensured that previous versions of stored data structures are
      properly migrated to new formats.

### API Changes

- [ ] Server - The server API is backwards-compatible
- The following client APIs are backwards-compatible:
  - [x] C++
  - [x] TypeScript
  - [x] Python

### Breaking Changes

If anything in this section is not true, please list all breaking changes.

### CODE REVIEW

File: pluto/src/vis/schematic/Forms.tsx [hunk 1]
```diff
@@ -672,6 +659,12 @@ export const ButtonTelemForm = ({ path }: { path: string }): ReactElement => {
       <Input.Item label="Output Channel" grow>
         <Channel.SelectSingle value={sink.channel} onChange={handleSinkChange} />
       </Input.Item>
+      <Form.NumericField
+        label="Delay"
```
Comment (pjdotson) [inline pluto/src/vis/schematic/Forms.tsx:663]: What do you think about calling it "Activation Delay" or "Hold Threshold" instead? I don't think "Delay" is descriptive enough, it wouldn't be immediately obvious to somebody that is playing around with a schematic and plops down a button. We could also add a tooltip when hovering over the label that gives a description that says "How long the button must be held before the command is sent" or something similar.

**Review by pjdotson — CHANGES_REQUESTED** [review]
1. When I press the button with a delay on it, the text inside moves slightly.

https://github.com/user-attachments/assets/e5c054d9-3d26-4862-af8d-6421a736a986

2. When pressed, the in-progress bar is not the same width and height as the button, which means there is a thin blue outline.

image

**Comment by emilbon99** [comment]
1. The inside text moving slightly is a webkit aberration on MacOS. I've tried to fix it many times with no luck.
2. The thin outline was intentional - I thought it looked nice. If you want it changed, I'm ok with doing so.

File: x/ts/src/telem/telem.ts [hunk 2]
```diff
@@ -470,6 +470,22 @@ export class TimeSpan implements Stringer {
     this.value = BigInt(value.valueOf());
   }
 
+  static fromSeconds(span: CrudeTimeSpan): TimeSpan {
+    if (span instanceof TimeSpan) return span;
+    if (span instanceof Rate) return span.period;
```
[the remark this answers is not in the document]
Comment (emilbon99) [reply x/ts/src/telem/telem.ts:475]: They should be

File: x/ts/src/telem/telem.ts [hunk 3]
```diff
@@ -470,6 +470,22 @@ export class TimeSpan implements Stringer {
...
+    if (["number", "bigint"].includes(typeof span))
+      return TimeSpan.seconds(span as number);
```
[the remark this answers is not in the document]
Comment (emilbon99) [reply x/ts/src/telem/telem.ts:477]: Fixed

File: x/ts/src/telem/telem.ts [hunk 4]
```diff
@@ -470,6 +470,22 @@ export class TimeSpan implements Stringer {
...
+    return new TimeSpan(span);
+  }
+
+  static fromMilliseconds(span: CrudeTimeSpan): TimeSpan {
+    if (span instanceof TimeSpan) return span;
+    if (span instanceof Rate) return span.period;
+    if (["number", "bigint"].includes(typeof span))
+      return TimeSpan.milliseconds(span as number);
```
[the remark this answers is not in the document]
Comment (emilbon99) [reply x/ts/src/telem/telem.ts:485]: Fixed

File: x/ts/src/telem/telem.ts [hunk 5]
```diff
@@ -470,6 +470,22 @@ export class TimeSpan implements Stringer {
...
 
+  static fromSeconds(span: CrudeTimeSpan): TimeSpan {
+    if (span instanceof TimeSpan) return span;
+    if (span instanceof Rate) return span.period;
+    if (["number", "bigint"].includes(typeof span))
+      return TimeSpan.seconds(span as number);
+    return new TimeSpan(span);
+  }
+
+  static fromMilliseconds(span: CrudeTimeSpan): TimeSpan {
```
[the remark this answers is not in the document]
Comment (emilbon99) [reply x/ts/src/telem/telem.ts:481]: Done

File: x/ts/src/telem/telem.ts [hunk 6]
```diff
@@ -470,6 +470,22 @@ export class TimeSpan implements Stringer {
...
+    if (span instanceof TimeSpan) return span;
+    if (span instanceof Rate) return span.period;
```
[the remark this answers is not in the document]
Comment (emilbon99) [reply x/ts/src/telem/telem.ts:483]: Added to CrudeTimeSpan

File: x/ts/src/telem/telem.ts [hunk 7]
```diff
@@ -470,6 +470,22 @@ export class TimeSpan implements Stringer {
     this.value = BigInt(value.valueOf());
   }
 
+  static fromSeconds(span: CrudeTimeSpan): TimeSpan {
```
[the remark this answers is not in the document]
Comment (emilbon99) [reply x/ts/src/telem/telem.ts:473]: Added

File: pluto/src/vis/schematic/Forms.tsx (hunk 1 above)
Comment (emilbon99) [reply pluto/src/vis/schematic/Forms.tsx:663]: Renamed to activation delay. We can add those tooltip helpers in a separate issue

File: x/ts/src/spatial/bounds/bounds.ts [hunk 8]
```diff
@@ -477,7 +477,7 @@ export const buildInsertionPlan = <T extends numeric.Value>(
  * @throws {Error} If invalid bounds are provided, such as bounds arrays not being of
  * length 2.
  *
- * @see {@link construct} for constructing valid bounds.
+ * @link {@link construct} for constructing valid bounds.
```
Comment (pjdotson) [inline x/ts/src/spatial/bounds/bounds.ts:480]: These two diffs are a bit odd, why is `@link` repeated?

File: client/ts/src/index.ts [hunk 9]
```diff
@@ -60,5 +59,7 @@ export {
   type TypedArray,
   type TZInfo,
 } from "@synnaxlabs/x/telem";
-import { control } from "@synnaxlabs/x";
+import { control, math } from "@synnaxlabs/x";
 export const Authority = control.Authority;
+/** @deprecated here for backwards compatibility. Use @synnaxlabs/x's math.Numeric instead. */
+export type NumericTelemValue = math.Numeric;
```
Comment (pjdotson) [inline client/ts/src/index.ts:65]: Do we actually think anyone is using this? I think we should just remove the export in general.

**Comment by pjdotson** [comment]
> 1. The inside text moving slightly is a webkit aberration on MacOS. I've tried to fix it many times with no luck.
> 2. The thin outline was intentional - I thought it looked nice. If you want it changed, I'm ok with doing so.

1. Okay, can you create an issue and mark it as low priority then?
2. Personally I think it would look better without the outline, but I don't feel strongly about it.

**Review by pjdotson — COMMENTED** [review]
Approve after comments get addressed and all checks pass.

**Comment by emilbon99** [comment]
Turns out fixing the outline is kind of hard - don't think it's worth the effort rn.

File: x/ts/src/spatial/bounds/bounds.ts (hunk 8 above)
Comment (pjdotson) [reply x/ts/src/spatial/bounds/bounds.ts:480]: @emilbon99 You only fixed one of them.

**Review by pjdotson — APPROVED** [review]
Approved after you address the last comment.

**Pull request merged**